ext/ManoptJuMPExt.jl

```diff
@@ -0,0 +1,408 @@
+module ManoptJuMPExt
+
+using Manopt
+using LinearAlgebra
+if isdefined(Base, :get_extension)
+    using JuMP: JuMP
+else
+    # imports need to be relative for Requires.jl-based workflows:
+    # https://github.com/JuliaArrays/ArrayInterface.jl/pull/387
+    using ..JuMP: JuMP
+end
+const MOI = JuMP.MOI
+using ManifoldsBase
+using ManifoldDiff
+
+function __init__()
+    # So that the user can use the convenient `Manopt.JuMP_Optimizer`
+    if isdefined(Base, :setglobal!)
+        setglobal!(Manopt, :JuMP_Optimizer, Optimizer)
+        setglobal!(Manopt, :JuMP_VectorizedManifold, VectorizedManifold)
+        setglobal!(Manopt, :JuMP_ArrayShape, ArrayShape)
+    else
+        Manopt.eval(:(const JuMP_Optimizer = $Optimizer))
+        Manopt.eval(:(const JuMP_VectorizedManifold = $VectorizedManifold))
+        Manopt.eval(:(const JuMP_ArrayShape = $ArrayShape))
+    end
+    return nothing
+end
+
+struct VectorizedManifold{M<:ManifoldsBase.AbstractManifold} <: MOI.AbstractVectorSet
+    manifold::M
+end
+
+"""
+    MOI.dimension(set::VectorizedManifold)
+
+Return the representation side of points on the (vectorized in representation) manifold.
+As the MOI variables are real, this means if the [`representation_size`](https://juliamanifolds.github.io/ManifoldsBase.jl/stable/functions/#ManifoldsBase.representation_size-Tuple{AbstractManifold}) yields (in product) `n`, this refers to the vectorized point / tangent vector  from (a subset of ``\\mathbb R^n``).
+"""
+function MOI.dimension(set::VectorizedManifold)
+    return prod(ManifoldsBase.representation_size(set.manifold))
+end
+
+mutable struct Optimizer <: MOI.AbstractOptimizer
+    # Manifold in which all the decision variables leave
+    manifold::Union{Nothing,ManifoldsBase.AbstractManifold}
+    # Description of the problem in Manopt
+    problem::Union{Nothing,Manopt.AbstractManoptProblem}
+    # State of the optimizer
+    state::Union{Nothing,Manopt.AbstractManoptSolverState}
+    # Starting value for each variable
+    variable_primal_start::Vector{Union{Nothing,Float64}}
+    # Sense of the optimization, e.g., min, max or no objective
+    sense::MOI.OptimizationSense
+    # Model used to compute gradient of the objective function with AD
+    nlp_model::MOI.Nonlinear.Model
+    # Solver parameters set with `MOI.RawOptimizerAttribute`
+    options::Dict{String,Any}
+    function Optimizer()
+        return new(
+            nothing,
+            nothing,
+            nothing,
+            Union{Nothing,Float64}[],
+            MOI.FEASIBILITY_SENSE,
+            MOI.Nonlinear.Model(),
+            Dict{String,Any}(DESCENT_STATE_TYPE => Manopt.GradientDescentState),
+        )
+    end
+end
+
+MOI.get(::Optimizer, ::MOI.SolverVersion) = "0.4.37"
```

Comment:
Could we document this? It seems to be just a small helper for something, but if this is added in Manopt I would like to have it documented, just to see whether I have to get in touch with JuMP developers or can resolve it myself.

Reply:
Maybe we could set this to the newest one when we merge this, but besides that this is resolved :)